mlir_graphblas/engine.py

```diff
@@ -762,6 +742,14 @@ def add(
 
         function_names: List[str] = []
         mlir_ast = parse_mlir_string(mlir_text)
+        # Find the inner-most Module
```

Comment:
I think using the inner-most module is not the best approach. It would miss `function_1` in this example:
```
module  {
  func @function_1() {
    return
  }
  module  {
    func @function_2() {
      return
    }
  }
}
```

Perhaps a better approach would be to traverse `mlir_ast` and append onto `mlir_functions` every time we see a function. `_resolve_type_aliases` shows how to traverse a PyMLIR AST. 

This would also handled weirder edge cases like this one (which passes through `mlir-opt` without errors):
```
module  {
  func @function_1() {
    module  {
      func @function_weird() {
        return
      }
    }
    return
  }
  module  {
    func @function_2() {
      return
    }
  }
}
```